### ISSUE TITLE
DataGrid EditTemplate get row data

### ISSUE THREAD

**Issue by njannink** [stsrki/Blazorise#852]
At the moment it is only possible to get the edited value through the context in the edit template of a datagrid column. But in some cases I need the entire object to be able to show the right editor and or disable something. Now I have to make a dummy property on the Model that exposes the entire object.

Would be nice if next to `@context` there would also be something like `@rowdata` or `@model` accessible in the EditTemplate

**Comment by AntonTurba** [comment]
I second @njannink's request.  I've been trying to throw together a drop-down column for a prototype Blazor project at work. 
Since I have the drop-down setup inside the EditTemplate, it would be preferable to also store the state/current selected value inside my row object.

**Comment by Luk164** [comment]
@stsrki I have encountered a similar problem. We already have `@context` that is made from a specified field of a object, can the object itself be made available? Like for example the `@model` @njannink mentioned?

**Comment by David-Moreira** [comment]
+1
Have also encountered this situation. Also ended up going with some workaround to grab the data, but seems like it would make sense to have the row/record data available to do some column logic based on the entire row/record.
It seems to me, that each one of the Template's provided in the DataGridColumn, has a different "behaviour". One is exposing a context, another a record, etc...

![image](https://user-images.githubusercontent.com/22283161/99151670-b9d9b500-2694-11eb-9be4-0679385f706d.png)

So a suggestion: they could be normalized behaving the same, exposing their own context with the row record available.
DisplayTemplate -> Display**Context** (provide row record)
FilterTemplate -> Filter**Context** (provide row record)
EditTemplate -> Edit**Context** (provide row record)
CaptionTemplate -> Caption**Context** (provide row record)

**Comment by Luk164** [comment]
@David-Moreira That looks even better. The PR I made is much simpler because I was afraid to make any breaking changes, so I just added the row record value into the `CellEditContext`. BTW the behaviour you proposed is pretty much the same as the WPF and UWP systems use for their datagrids.

**Comment by David-Moreira** [comment]
Nice, did not know that was the system on those data grids, I'm more of a web kinda guy! 
And It's not like everything is created perfectly the first time.
Sometimes breaking changes are just needed. 😆
Let's see what's @stsrki take on this.

**Comment by stsrki** [comment]
@Luk164 I'm, not sure about this PR. The problem is that having access to the edit model there are side-effects. If you change anything in the model it cannot be reversed. So, you go into row edit mode, you start changing fields, and then click **cancel**. If you changed anything in the model directly that will stay changed. And that is precisely what I wanted to avoid with `CellEditContext`. The only way around it is to create a deep copy of the model that doesn't have any **reference** to the original model. This is something I have in mind for a long time but still didn't come(or chose) with the best solution.

@David-Moreira Every template has a different behavior because when you look at them all separately it makes sense to make them behave differently.

- CaptionTemplate is responsible for the column header so it cannot have any row model. There can be a thousand rows in the grid but only one header.
- EditTemplate cannot have access to the model as that can lead to the problems I mentioned above.
- FilterTemplate the same as CaptionTemplate
- DisplayTemplate already has access to the model as it's the only one that should work with the model directly.

While I can see a reason for exposing the model with EditTemplate it is not so easy. And you can always get the model by using the `SelectedRow` handler.

**Comment by Luk164** [comment]
@stsrki I see your point, but not having access to the model just because of fear that the programmer might mess up is not the right approach. Instead it should be clearly documented so that it is obvious that changes to `cellEditContext` are reversible while changes the `model` are instant and thus discouraged. We need the model for decision-making anyways. For example I have an object with three modes of operation, each mode requires me to sow different kinds of edits. What I want is to to hide/show edit fields in real-time once user changes the mode. Also deep cloning is pretty much useless for complex objects like I use.

TLDR: It is never a good idea to limit functionality just because a programmer might mess up. I believe my approach would enable all the functionality with minimal danger of side effects.

**Comment by David-Moreira** [comment]
@stsrki Oh, what I meant by different behaviour, is they expose "different "type" of APIs", 
Why wouldn't they all be "normalized" and expose a **context** API? Each with their own properties & behaviour of course.
For example, if you were now to expose something new, like a property to the DisplayTemplate or Caption Template, you would have to make a breaking change. But if they already exposed a **context** API, you would only be adding the new property to their own exposed context.

@Luk164 
I would have to go with @stsrki on this one, there are certain internal details that shouldn't be exposed in order to avoid bugs and guarantee the correct behaviour of the exposed API. 
@stsrki 
While to be fair, I'm not sure if the exposed SelectedRow can be mutated and mess up with the Edit behaviour all the same.

**Comment by Luk164** [comment]
@David-Moreira What I did is pass the model for creating a row together with `CellEditContext` Value. This means ANY changes to the model will take effect immediately. My point is that having the ability to affect your model instead of just reading it can allow for incorrect use - where clicking cancel instead of save will not fully revert changes. I fully believe this tiny drawback is acceptable for the gain of access to the model for all kinds of purposes, and even the direct modification could be useful in some specific cases. I also believe the drawback can be mitigated by proper documentation.

**Comment by stsrki** [comment]
@David-Moreira Well, `context` API or keyword comes with every render-fragment. It's the default keyword by Blazor, and it's not specific to Blazorise. So we already have that. If I understood you correctly... 😅

Regarding the SelectedRow. It's true that in the end, it can lead to the same problem as in Edit template, but at least a developer is aware that he/she is bending the rules slightly. So in case something breaks, it's more obvious.

@Luk164 Having a warning in the documentation is one way of doing it. But not a lot of people read the documentation and even in the case they do, it can be overlooked easy.

**Comment by Luk164** [comment]
@stsrki I agree (I am also guilty of this), but the mistake is quite obvious and should not be that hard to pick up on. Also it is quite intuitive that modifying `model` will actually modify model. 

One more thing we could do is add descriptive comments that would help most IDEs show the purpose of the model reference. If anyone ignores both documentation, comments, looking up the known possible issue AND cannot find out why modifications are sticking by himself, well, we can't help them all can we? I stand behind my proposal because I believe it is the best compromise we can make and the gains outweigh the one very specific drawback.

**Comment by David-Moreira** [comment]
@stsrki  maybe I'm just being picky, this is what I meant:

![image](https://user-images.githubusercontent.com/22283161/99255679-32af4d00-280c-11eb-9b4d-1bf15199d708.png)

**Comment by stsrki** [comment]
@David-Moreira Ah I understand now. 🤔 Yeah, it can be useful. Since it's a breaking change I will need to consider it before trying to implement it.

**Comment by David-Moreira** [comment]
@stsrki Yep, unfortunately now it's a breaking change. 👍  
Just that EditTemplate and FilterTemplate, already work like that.
Maybe in some of the next major releases, or when the actual need to add more properties/behaviour arises. :D

**Comment by Luk164** [comment]
@stsrki I have an idea, if we add an optional parameter to `EditContext`, we can use it to show/hide the model inside, something like `ExposeUnderlyingContext=true`. That way we can almost guarantee `model` will not be misused.

**Comment by stsrki** [comment]
Context doesn't work like that, unfortunately. A model can be either defined or undefined before EditTemplate is activated.

**Comment by Luk164** [comment]
@stsrki You could make an accessor that throws if you use it w/o enabling it.

**Comment by njannink** [comment]
Now we have C# 9 with records which is something you ideally use in things like these DataGrids you will have immutable objects. So editing the row-data by both the user and also the Datagrid behavior upon apply is something that won't be possible anymore anyway. So upon apply I would expect to get the original row object and a list of changes so you (or better your backend) can compile a new object that will be replaced in the collection.

**Comment by Luk164** [comment]
@njannink Good point, but EFCore does not work well with records yet.

**Comment by njannink** [comment]
I now do this by implementing the RowUpdating event and always send Cancel=true. And @Luk164 my rest API returns records, but the server backend uses its own EFCore datamodels. Automapper does the trick there

**Comment by Luk164** [comment]
@njannink I was just pointing out that it would not work for every application. I use blazorSS, so no API in the middle, just direct access.

**Comment by njannink** [comment]
Yes true. But a lot of microservice architectures use [CQRS](https://martinfowler.com/bliki/CQRS.html#:~:text=CQRS%20stands%20for%20Command%20Query,you%20use%20to%20read%20information.) or something similar. Which means you will never edit your Query results (records), but you will send Commands to your rest API to update your records. But this discussion is derailing the original topic. I just wanted to point out to @stsrki that updating directly the row-data upon apply might not be possible for everybody and that people might need to use workarounds to get the desired behavior. I (ab)use the `RowUpdating` event, but you could also make another abstraction (ViewModels) around the orinal Query results

**Comment by Luk164** [comment]
@njannink Right, back to topic at hand. We need some way to get `rowModel` in EditTemplate. Apart from my PR, the only other way I see is to make it available as differently named context in `DataGridRow` itself (probably it's base class actually but let's not get into details). @stsrki Already mentioned deep-copy, but I do not believe that is gonna work for anything complex. Does anyone have an alternative?

**Comment by stsrki** [comment]
Model added to the CellEditContext
